extensions/amp-fx-collection/0.1/amp-fx-collection.js

```diff
@@ -0,0 +1,175 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpEvents} from '../../../src/amp-events';
+import {listen} from '../../../src/event-helper';
+import {iterateCursor} from '../../../src/dom';
+import {dev, user, rethrowAsync} from '../../../src/log';
+import {map} from '../../../src/utils/object';
+import {Services} from '../../../src/services';
+import {ParallaxProvider} from './providers/parallax';
+
+const TAG = 'amp-fx-collection';
+
+/**
+ * Enum for list of supported visual effects.
+ * @enum {string}
+ */
+const FxType = {
+  PARALLAX: 'parallax',
+};
+
+/**
+ * Map of fx type to fx provider class.
+ * @type {Object<FxType, function(new:FxProviderInterface, !../../../src/service/ampdoc-impl.AmpDoc)>}
+ */
+const fxProviders = map();
+fxProviders[FxType.PARALLAX] = ParallaxProvider;
+
+/**
+ * Bootstraps elements that have `amp-fx=<fx1 fx2>` attribute and installs
+ * the specified effects on them.
+ */
+class AmpFxCollection {
+
+  /**
+   * @param  {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @private @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const {!Document|!ShadowRoot} */
+    this.root_ = ampdoc.getRootNode();
+
+    /** @private @const {!Array<!Element>} */
+    this.seen_ = [];
+
+    /** @private @const {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(ampdoc);
+
+    /** @private @const {!Object<FxType, FxProviderInterface>} */
+    this.fxProviderInstances_ = map();
+
+    ampdoc.whenReady().then(() => this.viewer_.whenFirstVisible()).then(() => {
```

Comment:
Is the `whenReady` necessary? Does the viewer ever resolve `whenFirstVisible` before `whenReady` is resolved? If either one could happen first, maybe something like this would better communicate your intent.

```js
Promise.all([
  ampdoc.whenReady(),
  this.viewer_.whenFirstVisible(),
]).then(/*...*/)
```

Reply:
yeah `whenFirstVisible` is purely based on a message received from the viewer so I assumed it can happen at any moment. Switched to `Promise.all`